File: Java/ObjekMati.cpp
#include <bits/stdc++.h>
#include "ObjekMati.hpp"
#include "Point.hpp"
#include "Aquarium.hpp"

int ObjekMati::counter = 0;

ObjekMati::ObjekMati(double newSpeed) : posisi(){
    id = counter++;
    jenis = "undefine";
    speed = newSpeed;
}

bool ObjekMati::operator==(const ObjekMati& other){
    return id == other.id;
}

bool ObjekMati::operator!=(const ObjekMati& other){
    return id != other.id;
}

int ObjekMati::getId(){
    return id;
}

string ObjekMati::getJenis(){
    return jenis;
}

Point ObjekMati::getPosisi(){
    return posisi;
}

double ObjekMati::getSpeed(){
    return speed;
}

void ObjekMati::setId(int newId){
    id = newId;
}

void ObjekMati::setJenis(string newJenis){
    jenis = newJenis;
}

void ObjekMati::setPosisi(Point newPosisi){
    posisi = newPosisi;
}

void ObjekMati::setSpeed(double newSpeed){
    speed = newSpeed;
}

bool ObjekMati::isDasar(Aquarium& aq){
    return this->getPosisi().getOrdinat() >= 600;
}

void ObjekMati::turun(Aquarium& aq){
    if (!isDasar(aq)){
      Point point = this->getPosisi();
      point.setOrdinat(this->getPosisi().getOrdinat() + this->getSpeed());
      this->setPosisi(point);
    } else {
      if (this->getJenis() == "Makanan Ikan"){
        aq.RemoveObject(this);
      }
    }
}

File: Java/ObjekMati.java
public class ObjekMati implements Comparable{

  protected static int counter = 0;
  protected final int id;
  protected int kecepatan;
  protected Point posisi;
  protected String jenis;

  public ObjekMati(int kecepatan) {
    this.id = counter++;
    this.jenis = null;
    this.kecepatan = kecepatan;
    this.posisi = new Point();
  }

  public int getId() {
    return id;
  }

  public int getKecepatan() {
    return kecepatan;
  }

  public void setKecepatan(int kecepatan) {
    this.kecepatan = kecepatan;
  }

  public String getJenis() {
    return jenis;
  }

  public void setJenis(String jenis) {
    this.jenis = jenis;
  }

  public Point getPosisi() {
    return posisi;
  }

  public void setPosisi(Point posisi) {
    this.posisi = posisi;
  }

  public boolean isDasar(Aquarium aquarium) {
    return this.getPosisi().getOrdinat() >= aquarium.getLength();
  }

  public void turun(Aquarium aquarium) {
    if (!isDasar(aquarium)) {
      Point point = this.getPosisi();
      point.setOrdinat(this.getPosisi().getOrdinat() + this.getKecepatan());
      this.setPosisi(point);
    } else {
      if (this.getJenis() == "Makanan Ikan") {
        aquarium.removeObject(this);
      }
    }
  }

  public int compareTo(Object objekMati){
    ObjekMati objek = (ObjekMati) objekMati;
    if(objek.getId() == id){
      return 0;
    }
    else{
      return 1;
    }
  }
}
